### ISSUE TITLE
Question about result of MergeJoin

### ISSUE THREAD

**Issue by REN-I-9jo9** [roadrunnerlenny/etlbox#55]
Sorry my English is poor,it might look strange.
I was try to use MergeJoin to join two table, but the result was weird.
Here's my testing code
```
int i = 0, j = 0;
            var source1 = new CustomSource<string>(()=> {
                i++;
                return $"s1_{i}";
            },()=> {
                return i >= 3;
            });
            var source2 = new CustomSource<string>(() => {
                j++;
                return $"s2_{j}";
            }, () => {
                return j >= 3;
            });
            
            MergeJoin<string, string, string> trans = new MergeJoin<string, string, string>(
                (s1,s2) =>
                {
                    return s1 + " " + s2;
                }
            );

            BlockTransformation<string> block = new BlockTransformation<string>(
            inputData => {
                return inputData;
            });
            
            var result = new List<string>();
            var dest = new CustomDestination<string>(
                (v)=> {
                    result.Add(v);
                });
            
            source1.LinkTo(trans.Target1);
            source2.LinkTo(trans.Target2);
            trans.LinkTo(block);
            block.LinkTo(dest);
            
            source1.Execute();
            source2.Execute();
            dest.Wait();

            foreach (var r in result)
                Console.WriteLine(r);
```

I thought result should be
```
s1_1 s2_1
s1_1 s2_2
s1_1 s2_3
s1_2 s2_1
s1_2 s2_2
s1_2 s2_3
s1_3 s2_1
s1_3 s2_2
s1_3 s2_3
```
but it was
```
s1_1 s2_1
s1_2 s2_2
s1_3 s2_3
```

and the MergeJoin also have no place to set the join condition.
Did I miss anything?

**Comment by roadrunnerlenny** [comment]
Hi @REN-I-9jo9 ,
thanks for trying out ETLBox and your feedback. 

The result you actually get from the merge join is the expected result. Perhaps the documentation is weak here, I will have a look at this and try to improve it. The idea of a merge join is to combine the incoming rows with each other by a given join function. If you need to filter out data, you can use predicates in the LinkTo method. But there is now way of getting more rows out than the MergeJoin receives. The MergeJoin works best with already sorted inputs. The Merge join is a "non-blocking" transformation, which means that it does only store the current processing row in memory (plus a little bit more for buffering + performance)

The result you wanted to see from the MergeJoin could be better described as a Cartesian product or the result of a Cross join. Basically, you have two tables: A und B, and every row in A should join with each row in B. This can not be accomplished with the current implementation of MergeJoin. The reason for this is that a cross join is a very memory-intense function, as you need to store at least one table in whole in the memory. This would make the MergeJoin a "Blocking" transformation, as it would need to store all data from one table in the memory first before it could continue processing.

This behavior is similar to how the LookupTransformation works. The lookup has a source, loads the data from the source into the memory and then applies some logic to every incoming row - while it has access to all data from the source. 
So you probably could create a similar behavior with a modified LookupTransformation. 

I can see that the result of a cross join is a use-case I haven't thought about yet, but can be very useful. So if you give me some days, I can create an implementation for a cross join  in ETLBox. This will probably be a new object (e.g. `CrossJoin`) that could be used instead of the MergeJoin.

**Comment by REN-I-9jo9** [comment]
Thanks!
I will wait for the update.
Can you also add the join condition to `CrossJoin`? It will be helpful.

**Comment by roadrunnerlenny** [comment]
Hi,
sure, I will try to. Though it‘s not a real join condition, as more a concatenation function. So basically a function that describe the result of the cross join. But it would be possible to return null, which actually would filter the record out.

If you are looking for a real join (left oder inner), you should look at the LookupTransformation. This result is basically equivalent to a full inner or left join.

**Comment by roadrunnerlenny** [comment]
Hi @REN-I-9jo9 ,

i just released version 1.8.2 (https://www.nuget.org/packages/ETLBox/1.8.2)
It now contains a CrossJoin. Here is an example how to use it:

```
            MemorySource<string> source1 = new MemorySource<string>();
            source1.Data = new List<string>() { "A", "B" };
            MemorySource<int> source2 = new MemorySource<int>();
            source2.Data = new List<int>() { 1, 2, 3 };
            CrossJoin<string, int, string> crossJoin = new CrossJoin<string, int, string>(
                (data1, data2) => data1 + data2.ToString()
            );
            MemoryDestination<string> dest = new MemoryDestination<string>();

            source1.LinkTo(crossJoin.InMemoryTarget);
            source2.LinkTo(crossJoin.PassingTarget);
            crossJoin.LinkTo(dest);
            source1.Execute();
            source2.Execute();
            dest.Wait();
```

The MemoryDestination will contain (in `dest.Data`) a List<string> that contains A1,B1,A2,B2,C1,C2.

Please note that this transformation is "partially blocking." Everything that goes into the InMemoryTarget  needs to be loaded into the memory first before the processing of the cross join can start. Therefore the InMemory-Target always should contain the smaller amount of data.

**Comment by REN-I-9jo9** [comment]
Thanks,It works.
